lib/backend/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,325 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/gravitational/teleport/lib/backend"
+	kubeutils "github.com/gravitational/teleport/lib/kube/utils"
+	"github.com/gravitational/trace"
+	corev1 "k8s.io/api/core/v1"
+	kubeerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applyconfigv1 "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	secretIdentifierName   = "state"
+	namespaceEnv           = "KUBE_NAMESPACE"
+	teleportReplicaNameEnv = "TELEPORT_REPLICA_NAME"
+	releaseNameEnv         = "RELEASE_NAME"
+)
+
+// InKubeCluster detemines if the agent is running inside a Kubernetes cluster and has access to
+// service account token and cluster CA. Besides, it also validates the presence of `KUBE_NAMESPACE`
+// and `TELEPORT_REPLICA_NAME` environment variables to generate the secret name.
+func InKubeCluster() bool {
+	_, _, err := kubeutils.GetKubeClient("")
+
+	return err == nil &&
+		len(os.Getenv(namespaceEnv)) > 0 &&
+		len(os.Getenv(teleportReplicaNameEnv)) > 0
+}
+
+// Config structure represents configuration section
+type Config struct {
+	// Namespace is the Agent's namespace
+	// Field is required
+	Namespace string
+	// SecretName is unique secret per agent where state and identity will be stored.
+	// Field is required
+	SecretName string
+	// ReplicaName is the Agent's pod name
+	// Field is required
+	ReplicaName string
+	// ReleaseName is the HELM release name
+	// Field is optional
+	ReleaseName string
+	// KubeClient is the Kubernetes rest client
+	// Field is required
+	KubeClient kubernetes.Interface
+}
+
+// Backend uses Kubernetes Secrets to store identities.
+type Backend struct {
+	// kubernetes client
+	k8sClientSet kubernetes.Interface
+	namespace    string
+	secretName   string
+	replicaName  string
+	releaseName  string
+
+	// Mutex is used to limit the number of concurrent operations per agent to 1 so we do not need
+	// to handle retries locally.
+	// The same happens with SQlite backend.
+	mu *sync.Mutex
```

Comment:
nit: `Backend` is heap allocated and doesn't move, so this can be a `sync.Mutex` by value rather than a pointer.